rcl_action/include/rcl_action/action_server.h

```diff
@@ -0,0 +1,727 @@
+ * \param[in] ros_result_response a ROS result response message to send
+ * \return `RCL_RET_OK` if the response was sent successfully, or
+ * \return `RCL_RET_INVALID_ARGUMENT` if any arguments are invalid, or
+ * \return `RCL_RET_ACTION_SERVER_INVALID` if the action server is invalid, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs. */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_send_result_response(
+  const rcl_action_server_t * action_server,
+  const void * ros_result_response);
+
+/// Take a pending cancel request using a rcl_action_server_t.
+/**
+ * \todo TODO(jacobperron) blocking of take?
+ *
+ * \todo TODO(jacobperron) pre-, during-, and post-conditions for message ownership?
+ *
+ * \todo TODO(jacobperron) is this thread-safe?
+ *
+ * The caller is responsible for ensuring that the type of `ros_cancel_request`_
+ * and the type associate with the client (via the type support) match.
+ * Passing a different type produces undefined behavior and cannot be checked
+ * by this function and therefore no deliberate error will occur.
+ *
+ * After receiving a successful cancel request, the appropriate goals can be
+ * transitioned to the state CANCELING using rcl_action_process_cancel_request().
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] action_server handle to the action server that will take the cancel request
+ * \param[out] ros_cancel_request a preallocated ROS cancel request where the request
+ *   message is copied
+ * \return `RCL_RET_OK` if the response was sent successfully, or
+ * \return `RCL_RET_INVALID_ARGUMENT` if any arguments are invalid, or
+ * \return `RCL_RET_ACTION_SERVER_INVALID` if the action server is invalid, or
+ * \return `RCL_RET_ACTION_SERVER_TAKE_FAILED` if take failed but no error occurred
+ *         in the middleware, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_take_cancel_request(
+  const rcl_action_server_t * action_server,
+  void * ros_cancel_request);
+
+/// Process a cancel request using a rcl_action_server_t.
+/**
+ * This is a non-blocking call.
+ *
+ * This function will transition one or more active goals to the state CANCELING.
+ * The following cancel policy applies based on the goal ID and the timestamp
+ * contained in the cancel request:
+ *
+ * - If the goal ID is zero and timestamp is zero, cancel all goals.
+ * - If the goal ID is zero and timestamp is not zero, cancel all goals accepted
+ *   at or before the timestamp.
+ * - If the goal ID is not zero and timestamp is zero, cancel the goal with the
+ *   given ID regardless of the time it was accepted.
+ * - If the goal ID is not zero and timestamp is not zero, cancel the goal with the
+ *   given ID and all goals accepted at or before the timestamp.
```

Comment:
@jacobperron would it make sense to expose cancelling operations on a per goal handle basis too and build `rcl_action_process_cancel_request()`  on top of them instead?

Reply:
The goal handle's state can be updated with [rcl_action_update_goal_state()](https://github.com/ros2/rcl/blob/48519db9c82a2b6f98b251779dd0732bef03a516/rcl_action/include/rcl_action/goal_handle.h#L194), which I plan to call from this  function.